### PR TITLE
Implement CloudFormation waiters

### PR DESCRIPTION
**Pull request by quiver**
# Cloudformation Waiters

Two sub-commands are added with this PR

## wait stack-available

to be used with `create-stack`/`update-stack`

    $ aws cloudformation create-stack --stack-name foo --template-body file://foo.template.json
    $ aws cloudformation wait stack-available --stack-name foo

o success status

* CREATE_COMPLETE
* UPDATE_COMPLETE

o retry status

* CREATE_IN_PROGRESS
* UPDATE_IN_PROGRESS
* UPDATE_COMPLETE_CLEANUP_IN_PROGRESS

o failure status

* other statuses

## wait stack-deleted

to be used with `delete-stack`

    $ aws cloudformation delete-stack --stack-name foo
    $ aws cloudformation wait stack-deleted --stack-name foo

o success status

* DELETE_COMPLETE (when queried with stack-id)
* ValidationError (when queried with stack-name)

o retry status

* DELETE_IN_PROGRESS

o failure status

* other statuses

## retry handling

Poll every 30 seconds for 120 times(60 minutes) with `describe-stacks` API and query the response with `Stacks[].StackStatus`.

When provisioning services like cloudformation or DB(RDS, RedShift, etc) or Elastic BeanStalk or changing magnetic device to SSD,  CFn takes a long time, so I set threshold times(60 mins) longer than other wait commands.

# StackStatus

## List of StackStatus

http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-describing-stacks.html

* CREATE_COMPLETE
* CREATE_IN_PROGRESS
* CREATE_FAILED
* DELETE_COMPLETE
* DELETE_FAILED
* DELETE_IN_PROGRESS
* ROLLBACK_COMPLETE
* ROLLBACK_FAILED
* ROLLBACK_IN_PROGRESS
* UPDATE_COMPLETE
* UPDATE_COMPLETE_CLEANUP_IN_PROGRESS
* UPDATE_IN_PROGRESS
* UPDATE_ROLLBACK_COMPLETE
* UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS
* UPDATE_ROLLBACK_FAILED
* UPDATE_ROLLBACK_IN_PROGRESS

## StackStatus state diagram

Based on the document and playing with CFn and advices from my colleagues, I created a StackStatus state diagram.
This diagram does *not* cover every single flow, but will cover most of the state changes(I think).

* Blue-filled statuses are success
* Green-filled statuses are retry
* Other statuses are failure

### create-stack 

create-stack

* If you pass `--disable-rollback`, rollback won't occur.
* `CREATE_IN_PROGRESS` stack can be deleted with `DeleteStack` API

### update-stack

update-stack

* `UPDATE_IN_PROGRESS` stack can be canceled(rollback) with `CancelUpdateStack` API
* There's no `--disable-rollback` option for update stack

### delete-stack
delete-stack

* There's no rollback or abort flow for `DeleteStack` API 
* `DescribeStacks`'s `StackName` parameter can take either `StackName` or `StackID`.
  For `DELETE_COMPLETE`-state stacks, `DescribeStacks`
  * returns `ValidationError` when queried with `StackName`
  * returns `DELETE_COMPLETE` status when queried with `StackID`

Original graphviz gist files can be found here https://gist.github.com/quiver/e5496a91deda73bc9f84

# Solves

* aws/aws-cli#895
* boto/boto3#84

### CODE REVIEW

**Comment by rayluo** [comment]
Thanks! We will verify this change.

**Comment by quiver** [comment]
> Thanks! We will verify this change.

Thx. I especially appreciate feedbacks from AWS CFn teams, in which I have limited expertise.

**Comment by rikf** [comment]
Any updates on this PR? Would be great to have this merged in.

**Comment by kapilt** [comment]
Would be super useful, there's a half-dozen scripts around that attempt to emulate this, having it directly in core on the cli would simplify things considerably.

**Comment by copumpkin** [comment]
:+1: I just implemented a worse version of this, as I'm sure dozens of people have done before me

File: botocore/data/cloudformation/2010-05-15/waiters-2.json [hunk 1]
```diff
@@ -0,0 +1,197 @@
+{
+    "version": 2,
+    "waiters": {
+        "StackAvailable": {
+            "delay": 30,
```
Comment (copumpkin) [inline botocore/data/cloudformation/2010-05-15/waiters-2.json]: What informs the 30 value you picked here?

Comment (quiver) [inline botocore/data/cloudformation/2010-05-15/waiters-2.json]: > What informs the 30 value you picked here?

Most Cloudformation APIs take many minutes to complete. For long-polling, 15 seconds are too frequent, 60 seconds are too long(FYW, Cloudfront and Redshift use 60 secs as polling intervals), and I thought 30 seconds are a good compromise.

**Comment by analytically** [comment]
:+1: I'm doing this manually now...

**Comment by pinfoldd** [comment]
+1 This is something we are having to do manually at the moment, this would be really useful to have.

**Comment by rafal-gancarz-opencredo** [comment]
:+1: we would love to use it on our project!

**Comment by copumpkin** [comment]
@rayluo is there anything blocking this from getting merged? It's the oldest open PR now and looks fairly uncontroversial.

**Comment by copumpkin** [comment]
Also cc @kyleknap who appears to be more active these days.

**Comment by webratz** [comment]
I also would like to see this getting merged, as it eliminates the need for many workarounds.

**Comment by johnny-die-tulpe** [comment]
+1

**Comment by jamesls** [comment]
Just a quick update.  The only thing we'll need to resolve is that it looks like some of the other SDKs pulled in a different version of waiters for this service from a PR to their repo: https://github.com/aws/aws-sdk-ruby/blob/master/aws-sdk-core/apis/cloudformation/2010-05-15/waiters-2.json

We'll need to consolidate both versions of this file.  Given there are existing ruby customers using their version of waiters, we'll likely need to rename some of the waiter names here.

**Comment by quiver** [comment]
@jamesls 

> We'll need to consolidate both versions of this file. Given there are existing ruby customers using their version of waiters, we'll likely need to rename some of the waiter names here.

Ruby's waiter have

* StackExists
* StackCreateComplete
* StackUpdateComplete
* StackDeleteComplete

and my waiter have

* StackAvailable
* StackDeleted

So my first task is to follow ruby API's names, i.e., 

* split StackAvailable into StackCreateComplete/StackUpdateComplete
* rename StackDeleted to StackDeleteComplete
* add new waiter StackExists(I'll just copy & paste)

Do I need to change this PR's delay/maxAttempts conditions to ruby's ?

In my opinion, ruby' s wait conditions are too short. For example, 

```
    "StackUpdateComplete": {
      "delay": 30,
      "operation": "DescribeStacks",
      "maxAttempts": 5,
```
This StackUpdateComplete API waits only for 2.5 minutes. 
I doubt they use this API in production environments.

**Comment by jamesls** [comment]
I think it's fine to increase the delay/maxAttempts.  Those are backwards compatible changes we can get ruby to adopt.

**Comment by analytically** [comment]
Any idea when this will land in boto3? 1.2.7?

**Comment by quiver** [comment]
Updated PR to follow ruby SDK API names.

sample JSON files to play with CFn from AWS CLI can be found here
https://gist.github.com/quiver/7366a7db185982bb60e5

**Comment by jamesls** [comment]
Awesome, thanks for the update to the PR.  I'll take a look at this today.

**Comment by analytically** [comment]
Would it be possible to get this merged?

**Comment by jamesls** [comment]
Thanks again @quiver for the pull request.  Sorry for the delay in getting this merged.  This will be part of the next CLI/botocore release.

**Pull request closed without merging**